sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,372 @@
+    this.flushLock = new Object();
+  }
+
+  /** Starts the flushing daemon thread if data_buffer_time_limit_ms is set. */
+  public void startFlushThread() {
+    if (timeLimit > 0 && this.flushFuture == null) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * Register the outbound data logical endpoint, returns the FnDataReceiver for processing the
+   * endpoint's outbound data.
+   */
+  public <T> FnDataReceiver<T> registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    if (outputDataReceivers.containsKey(pTransformId)) {
+      throw new IllegalStateException(
+          "Outbound data endpoint already registered for " + pTransformId);
+    }
+    Receiver<T> receiver = new Receiver<>(coder);
+    if (timeLimit > 0) {
+      outputDataReceivers.put(pTransformId, receiver);
+      return data -> {
+        checkFlushThreadException();
+        synchronized (flushLock) {
+          receiver.accept(data);
+          if (totalByteCounter.longValue() > sizeLimit) {
+            flush();
+          }
+        }
+      };
+    }
+    outputDataReceivers.put(pTransformId, receiver);
+    return receiver;
+  }
+
+  /**
+   * Register the outbound timers logical endpoint, returns the FnDataReceiver for processing the
+   * endpoint's outbound timers data.
+   */
+  public <T> FnDataReceiver<T> registerOutputTimersLocation(
+      String pTransformId, String timerFamilyId, Coder<T> coder) {
+    TimerEndpoint timerKey = new TimerEndpoint(pTransformId, timerFamilyId);
+    if (outputTimersReceivers.containsKey(timerKey)) {
+      throw new IllegalStateException(
+          "Outbound timers endpoint already registered for " + timerKey);
+    }
+    Receiver<T> receiver = new Receiver<>(coder);
+    if (timeLimit > 0) {
+      outputTimersReceivers.put(timerKey, receiver);
+      return timers -> {
+        checkFlushThreadException();
+        synchronized (flushLock) {
+          receiver.accept(timers);
+        }
+      };
+    }
+    outputTimersReceivers.put(timerKey, receiver);
+    return receiver;
+  }
+
+  public void flush() throws IOException {
+    Elements.Builder elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    if (totalByteCounter.longValue() == 0
+        && outputTimersReceivers.isEmpty()
+        && outputDataReceivers.isEmpty()) {
```

Comment:
I don't think the byte count will ever be non zero if nothing is registered.
```suggestion
    if (outputTimersReceivers.isEmpty()
        && outputDataReceivers.isEmpty()) {
```